libs/pika/async_cuda/include/pika/async_cuda/cuda_pool.hpp

```diff
@@ -0,0 +1,95 @@
+//  Copyright (c) 2021 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <pika/assert.hpp>
+#include <pika/async_cuda/cuda_stream.hpp>
+#include <pika/coroutines/thread_enums.hpp>
+#include <pika/datastructures/optional.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <iosfwd>
+#include <memory>
+#include <vector>
+
+namespace pika::cuda::experimental {
+    /// A pool of CUDA streams, used for scheduling work on a CUDA device.
+    ///
+    /// The pool initializes a set of CUDA streams on construction and provides
+    /// access to the streams in a round-robin fashion. The pool is movable and
+    /// copyable with reference semantics. Copies of a pool still refer to the
+    /// original pool of streams.
+    class cuda_pool
+    {
+    private:
+        struct streams_holder
+        {
+            std::size_t num_streams;
+            std::vector<cuda_stream> streams;
+            std::atomic<std::size_t> active_stream_index{0};
+
+            PIKA_EXPORT streams_holder(int device, std::size_t num_streams,
+                pika::threads::thread_priority);
+            streams_holder(streams_holder&&) = delete;
+            streams_holder(streams_holder const&) = delete;
+            streams_holder& operator=(streams_holder&&) = delete;
+            streams_holder& operator=(streams_holder const&) = delete;
+
+            PIKA_EXPORT cuda_stream const& get_next_stream();
+        };
+
+        struct pool_data
+        {
+            int device;
+            streams_holder normal_priority_streams;
+            streams_holder high_priority_streams;
+
+            PIKA_EXPORT pool_data(int device,
+                std::size_t num_normal_priority_streams,
+                std::size_t num_high_priority_streams);
+            pool_data(pool_data&&) = delete;
+            pool_data(pool_data const&) = delete;
+            pool_data& operator=(pool_data&&) = delete;
+            pool_data& operator=(pool_data const&) = delete;
+        };
+
+        std::shared_ptr<pool_data> data;
+
+    public:
+        PIKA_EXPORT explicit cuda_pool(int device = 0,
+            std::size_t num_normal_priority_streams = 64,
+            std::size_t num_high_priority_streams = 64);
```

Comment:
This seems to be a better default for DLA-Future, but is it a good default for everyone?
```suggestion
        PIKA_EXPORT explicit cuda_pool(int device = 0,
            std::size_t num_normal_priority_streams = 64,
            std::size_t num_high_priority_streams = 1);
```

Reply:
I'm keeping this as it is. It seems very algorithm-specific what number of streams is the best. At least 64/64 gives enough concurrency for submitted work. If we come to some more universal "best parameters" we can still update this. Also, they're only defaults after all.